engines/saga/image.cpp

```diff
@@ -96,26 +114,49 @@ bool SagaEngine::decodeBGImage(const ByteArray &imageData, ByteArray &outputBuff
 	readS.readUint16();
 	readS.readUint16();
 
-	outputBuffer.resize(hdr.width * hdr.height);
-
 	if (isAGA() || isECS()) {
-		int planePitch = (hdr.width + 15) & ~15;
-		int linePitch = isAGA() ? planePitch : (planePitch * 5 / 8);
 		unsigned bitnum = isAGA() ? 8 : 5;
 		int headerSize = 8 + (3 << bitnum);
 		const byte *RLE_data_ptr = &imageData.front() + headerSize;
 		size_t RLE_data_len = imageData.size() - headerSize;
 
-		if (RLE_data_len != (size_t) linePitch * hdr.height)
-			return false;
-
-		memset(outputBuffer.getBuffer(), 0, hdr.width * hdr.height);
-		for (int y = 0; y < hdr.height; y++)
-			for (int x = 0; x < hdr.width; x++)
-				for (unsigned bit = 0; bit < bitnum; bit++) {
-					int inXbit = x + bit * planePitch;
-					outputBuffer[y * hdr.width + x] |= ((RLE_data_ptr[y * linePitch + inXbit / 8] >> (7 - inXbit % 8)) & 1) << bit;
-				}
+		if (getFeatures() & GF_POWERPACK_GFX) {
+			int aligned_width = (hdr.width + 15) & ~15;
+			int pitch = isAGA() ? aligned_width : (aligned_width * 5 / 8);
+			if (RLE_data_len < 12) {
+				warning("Compressed size too short");
+				return false;
+			}
+			if (READ_BE_UINT32 (RLE_data_ptr) != RLE_data_len - 4) {
+				warning("Compressed size mismatch: %d vs %d", READ_BE_UINT32 (RLE_data_ptr), (int)RLE_data_len - 4);
+				return false;
+			}
+			if (READ_BE_UINT32 (RLE_data_ptr + 4) != MKTAG('P', 'A', 'C', 'K')) {
+				warning("Compressed magic mismatch: 0x%08x vs %08x", READ_BE_UINT32 (RLE_data_ptr + 4), MKTAG('P', 'A', 'C', 'K'));
+				return false;
+			}
+			uint32 uncompressed_len = 0;
+			byte *uncompressed = Common::PowerPackerStream::unpackBuffer(RLE_data_ptr + 4, RLE_data_len - 4, uncompressed_len);
+			if (uncompressed == nullptr || (int) uncompressed_len != pitch * hdr.height) {
+				warning("Uncompressed size mismatch: %d vs %d", uncompressed_len, pitch * hdr.height);
+				return false;
+			}
+			if (isAGA() && pitch == hdr.width) {
+				// TODO: Use some kind of move semantics
```

Comment:
Is this affecting anything regarding functionality?

Reply:
No, just performance